dartvlc/api/api.cc

```diff
@@ -245,123 +234,124 @@ DLLEXPORT void BroadcastCreate(int32_t id, const char* type,
   g_broadcasts->Get(id, std::move(media), std::move(configuration));
 }
 
-DLLEXPORT void BroadcastStart(int32_t id) {
+void BroadcastStart(int32_t id) {
   Broadcast* broadcast = g_broadcasts->Get(id, nullptr, nullptr);
   broadcast->Start();
 }
 
-DLLEXPORT void BroadcastDispose(int32_t id) { g_broadcasts->Dispose(id); }
+void BroadcastDispose(int32_t id) { g_broadcasts->Dispose(id); }
 
-DLLEXPORT void ChromecastCreate(int32_t id, const char* type,
-                                const char* resource, const char* ip_address) {
+void ChromecastCreate(int32_t id, const char* type, const char* resource,
+                      const char* ip_address) {
   std::shared_ptr<Media> media = Media::create(type, resource);
   chromecasts->Get(id, std::move(media), ip_address);
 }
 
-DLLEXPORT void ChromecastStart(int32_t id) {
+void ChromecastStart(int32_t id) {
   Chromecast* chromecast = chromecasts->Get(id, nullptr, "");
   chromecast->Start();
 }
 
-DLLEXPORT void ChromecastDispose(int32_t id) { chromecasts->Dispose(id); }
+void ChromecastDispose(int32_t id) { chromecasts->Dispose(id); }
 
-DLLEXPORT void RecordCreate(int32_t id, const char* saving_file,
-                            const char* type, const char* resource) {
+void RecordCreate(int32_t id, const char* saving_file, const char* type,
+                  const char* resource) {
   std::shared_ptr<Media> media = Media::create(type, resource);
   g_records->Get(id, media, saving_file);
 }
 
-DLLEXPORT void RecordStart(int32_t id) {
+void RecordStart(int32_t id) {
   Record* record = g_records->Get(id, nullptr, "");
   record->Start();
 }
 
-DLLEXPORT void RecordDispose(int32_t id) { g_records->Dispose(id); }
-
-DLLEXPORT char** DevicesAll() {
-  std::vector<Device> devices = Devices::All();
-  g_devices_ptr = new char*[(devices.size() * 2) + 1];
-  g_devices_size = (devices.size() * 2) + 1;
-  g_devices_ptr[0] = new char[200];
-  strncpy(g_devices_ptr[0], std::to_string(devices.size()).data(), 200);
-  int32_t index = 0;
-  for (Device& device : devices) {
-    g_devices_ptr[index + 1] = new char[200];
-    strncpy(g_devices_ptr[index + 1], device.id().data(), 200);
-    g_devices_ptr[index + 2] = new char[200];
-    strncpy(g_devices_ptr[index + 2], device.name().data(), 200);
-    index += 2;
-  }
-  return g_devices_ptr;
-}
+void RecordDispose(int32_t id) { g_records->Dispose(id); }
 
-DLLEXPORT void DevicesClear() {
-  if (g_devices_ptr != nullptr) {
-    for (size_t i = 0; i < g_devices_size; i++) {
-      delete g_devices_ptr[i];
-    }
-    delete[] g_devices_ptr;
-    g_devices_ptr = nullptr;
-    g_devices_size = 0;
-  }
+void DevicesClear(void*, void* peer) {
+  delete reinterpret_cast<std::vector<const char*>*>(peer);
 }
 
-DLLEXPORT char** EqualizerCreateEmpty() {
+struct DevicesStruct DevicesAll(Dart_Handle object) {
+  auto devices = new std::vector<Device>(Devices::All());
+  auto ids = new std::vector<const char*>();
```

Comment:
Previously I thought to use `char**` styled C array, but I believe using `std::vector` is a good idea:
- It won't need ask for size inside cleanup finalizer.
- It matches with rest of the code.
- This just requires devices vector to stay in heap along with these two `std::vector<char*>`, which we can easily wipe in finalizer.
- If I settled with `char**`, we'd end up with global variables again (which was our main motive to remove).